portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -13,20 +13,81 @@
 // limitations under the License.
 
 package com.google.sps.servlets;
+import com.google.sps.data.Comment;
+import com.google.sps.data.Constants;
+import com.google.gson.Gson;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.List;
+
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that handles comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
-  }
+    /** 
+     * Retrieves the (limit number) of most recent comments from Datastore
+     * and produces a JSON response containing these comments
+     */
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        int limit = Integer.parseInt(request.getParameter(Constants.LIMIT_PROPERTY));
+        Query query = new Query(Constants.COMMENT_KIND).addSort(Constants.TIMESTAMP_PROPERTY, SortDirection.DESCENDING);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
+
+        List<Comment> comments = new ArrayList<>();
+        int i = 0;
+        for (Entity entity : results.asIterable()) {
+            if (i == limit) {
+                break;
+            } 
+            i++;
+            long id = entity.getKey().getId();
+            String text = (String)entity.getProperty(Constants.TEXT_PROPERTY);
+            long timestamp = (long)entity.getProperty(Constants.TIMESTAMP_PROPERTY);
+
+            Comment comment = new Comment(id, text, timestamp);
+            comments.add(comment);
+        }
+
+        Gson gson = new Gson();
+
+        response.setContentType("application/json");
+        response.getWriter().println(gson.toJson(comments));
+    }
+
+    /** 
+     * Stores the comment into datastore and redirect user back to home page 
+     * after submission. datastore.put() directs to /data page so the redirect
+     * prevents the user from reaching the /data page.
```

Comment:
This is not accurate: as we discussed, `datastore.put()` is not the call responsible for the redirect. The reason we need a redirect is because we use a `form` in the HTML file.

Reply:
I changed the comment